x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/PackedValuesBlockHash.java

```diff
@@ -65,14 +65,14 @@ final class PackedValuesBlockHash extends BlockHash {
     private final BytesRefHash bytesRefHash;
     private final int nullTrackingBytes;
     private final BytesRefBuilder bytes = new BytesRefBuilder();
-    private final Group[] groups;
+    private final List<GroupSpec> specs;
```

Comment:
Ok. This was really fun to figure out. If `groups` lives on the object then the `add` and `lookup` methods aren't reentrant on any particular hash. And `lookup` can't be thread safe. Now, I don't try to use it in a multi-threaded way, but I *do* call `lookup` from inside of a callback for `add` - at least in tests. And they worked! But in a really muddled, sneaky, scary way. This has us building the `Group` objects on the fly. This is nicer because I can make them `Releasable`. It's more objects to build on add, but it's safer. Less crazy making.